### PR TITLE
harmony-one 4.2.1 (new formula)

### PR DESCRIPTION
**Pull request by cylim**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/harmony-one.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  url "https://github.com/harmony-one/harmony/archive/refs/tags/v4.2.1.tar.gz"
+  sha256 "d29609a99a3a2e9031d2a26414ffdc3e271c3bf9cf9331dfcb321715fef682d4"
+  license "LGPL-3.0-or-later"
+
+  depends_on "go" => :build
+
+  # arm64 issues: https://github.com/harmony-one/go-sdk/issues/271#issuecomment-932876822
+  depends_on arch: :x86_64
+  depends_on "gmp"
+  depends_on "openssl@1.1"
```
Comment (carlocab) [inline Formula/harmony-one.rb:13]: Can this use `openssl@3`?

File: Formula/harmony-one.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+  resource "go-sdk" do
+    url "https://github.com/harmony-one/go-sdk.git",
+      revision: "857736c1c51091d86049c3860604830903b046aa"
```
Comment (carlocab) [inline Formula/harmony-one.rb:17]:
```suggestion
        revision: "857736c1c51091d86049c3860604830903b046aa"
```

File: Formula/harmony-one.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+  resource "bls" do
+    url "https://github.com/harmony-one/bls.git",
+      revision: "2b7e49894c0f15f5c40cf74046505b7f74946e52"
```
Comment (carlocab) [inline Formula/harmony-one.rb:22]:
```suggestion
        revision: "2b7e49894c0f15f5c40cf74046505b7f74946e52"
```

File: Formula/harmony-one.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+  resource "mcl" do
+    url "https://github.com/harmony-one/mcl.git",
+      revision: "99e9aa76e84415e753956c618cbc662b2f373df1"
```
Comment (carlocab) [inline Formula/harmony-one.rb:27]:
```suggestion
        revision: "99e9aa76e84415e753956c618cbc662b2f373df1"
```

File: Formula/harmony-one.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  license "LGPL-3.0-or-later"
+
+  depends_on "go" => :build
+
+  # arm64 issues: https://github.com/harmony-one/go-sdk/issues/271#issuecomment-932876822
+  depends_on arch: :x86_64
+  depends_on "gmp"
+  depends_on "openssl@1.1"
+
+  resource "go-sdk" do
```
Comment (carlocab) [inline Formula/harmony-one.rb:15]: How are these resources updated? Please document this in comments above each resource.

File: Formula/harmony-one.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  def install
+    # See https://github.com/golang/go/issues/26487
+    ENV.O0 if OS.linux?
+
+    ENV["GOPATH"] = buildpath
+
+    dir = buildpath/"src/github.com/harmony-one/harmony"
+    dir.install buildpath.children
+
+    %w[mcl bls go-sdk].each do |r|
```
Comment (carlocab) [inline Formula/harmony-one.rb:39]: Does this not use Go modules?

File: Formula/harmony-one.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      end
+    end
+
+    cd buildpath/"src/github.com/harmony-one/go-sdk" do
+      system "make", "-j8"
+
+      lib.install Dir["../bls/lib/*.{dylib,so}"]
+      lib.install Dir["../mcl/lib/*.{dylib,so}"]
+      bin.install "dist/hmy"
+      prefix.install_metafiles
```
Comment (carlocab) [inline Formula/harmony-one.rb:54]:
```suggestion
```

File: Formula/harmony-one.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        system "make", "-j8", "BLS_SWAP_G=1" if r == "bls"
+      end
+    end
+
+    cd buildpath/"src/github.com/harmony-one/go-sdk" do
+      system "make", "-j8"
+
+      lib.install Dir["../bls/lib/*.{dylib,so}"]
+      lib.install Dir["../mcl/lib/*.{dylib,so}"]
+      bin.install "dist/hmy"
```
Comment (carlocab) [inline Formula/harmony-one.rb:53]: Is there no `make install`?

File: Formula/harmony-one.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+      temp_dir = buildpath/"src/github.com/harmony-one/#{r}"
+
+      resource(r).stage do
+        mv pwd, temp_dir
+        system "make", "-j8", "BLS_SWAP_G=1" if r == "bls"
+      end
+    end
+
+    cd buildpath/"src/github.com/harmony-one/go-sdk" do
+      system "make", "-j8"
```
Comment (carlocab) [inline Formula/harmony-one.rb:49]:
```suggestion
      system "make"
```

File: Formula/harmony-one.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+    dir = buildpath/"src/github.com/harmony-one/harmony"
+    dir.install buildpath.children
+
+    %w[mcl bls go-sdk].each do |r|
+      temp_dir = buildpath/"src/github.com/harmony-one/#{r}"
+
+      resource(r).stage do
+        mv pwd, temp_dir
+        system "make", "-j8", "BLS_SWAP_G=1" if r == "bls"
```
Comment (carlocab) [inline Formula/harmony-one.rb:44]: Aren't you calling `make` from a nonexistent directory here, given that you've `mv`ed it elsewhere?

Comment (cylim) [reply Formula/harmony-one.rb:44]: Moving the current directory to another directory, you will still in the directory.

File: Formula/harmony-one.rb (hunk 8 above)
Comment (cylim) [reply Formula/harmony-one.rb:53]: make included all the steps required to build.

File: Formula/harmony-one.rb (hunk 9 above)
Comment (cylim) [reply Formula/harmony-one.rb:49]: will remove it, thanks. :)

File: Formula/harmony-one.rb (hunk 2 above)
Comment (cylim) [reply Formula/harmony-one.rb:17]: Will add extra spacing, thanks 😄

File: Formula/harmony-one.rb (hunk 8 above)
Comment (carlocab) [reply Formula/harmony-one.rb:53]: Yes, but what about the steps required to install?

File: Formula/harmony-one.rb (hunk 2 above)
Comment (cylim) [reply Formula/harmony-one.rb]: added extra spacing.

File: Formula/harmony-one.rb (hunk 1 above)
Comment (cylim) [reply Formula/harmony-one.rb]: seems like it build successfully with `openssl@3`, I would check if anything broken internally.

File: Formula/harmony-one.rb (hunk 7 above)
Comment (cylim) [reply Formula/harmony-one.rb]: removed

File: Formula/harmony-one.rb (hunk 8 above)
Comment (cylim) [reply Formula/harmony-one.rb:53]: The steps required to install is in the Makefile, it is running a script on `harmony` repo for setting up and then running `go build` with go module turned on.

File: Formula/harmony-one.rb (hunk 5 above)
Comment (cylim) [reply Formula/harmony-one.rb:15]: harmony repo is where everything happening, whenever we need to upgrade the resources, we should add the linked repo in the changelog. Besides that, I will update the harmony doc for how to install with brew and how to upgrade the formula after this is approved.

**Pull request closed without merging**